plugins/generator-1.21.1/neoforge-1.21.1/procedures/projectiles_arrow.java.ftl

```diff
@@ -1,29 +1,17 @@
-<#assign projectile = generator.map(field$projectile, "projectiles", 1)>
-<#assign hasShooter = (input$shooter != "null")>
-<#assign isPiercing = (input$piercing != "/*@int*/0")>
-new Object() {
-	public Projectile getArrow(Level level<#if hasShooter>, Entity shooter</#if>, float damage, int knockback, byte piercing) {
-		AbstractArrow entityToSpawn = new ${generator.map(field$projectile, "projectiles", 0)}(${projectile}, level) {
-			@Override public byte getPierceLevel() {
-				return piercing;
-			}
-
-			@Override protected void doKnockback(LivingEntity livingEntity, DamageSource damageSource) {
-				if (knockback > 0) {
-					double d1 = Math.max(0.0, 1.0 - livingEntity.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
-					Vec3 vec3 = this.getDeltaMovement().multiply(1.0, 0.0, 1.0).normalize().scale(knockback * 0.6 * d1);
-					if (vec3.lengthSqr() > 0.0) {
-						livingEntity.push(vec3.x, 0.1, vec3.z);
-					}
-				}
-			}
-		};
-		<#if hasShooter>entityToSpawn.setOwner(shooter);</#if>
-		entityToSpawn.setBaseDamage(damage);
-		<#if field$projectile?starts_with("CUSTOM:")>entityToSpawn.setSilent(true);</#if>
-		<#if field$fire == "TRUE">entityToSpawn.igniteForSeconds(100);</#if>
-		<#if field$particles == "TRUE">entityToSpawn.setCritArrow(true);</#if>
-		<#if field$pickup != "DISALLOWED">entityToSpawn.pickup = AbstractArrow.Pickup.${field$pickup};</#if>
-		return entityToSpawn;
-	}
-}.getArrow(projectileLevel<#if hasShooter>, ${input$shooter}</#if>, ${opt.toFloat(input$damage)}, ${opt.toInt(input$knockback)}, (byte) ${input$piercing})
\ No newline at end of file
+<@addTemplate file="utils/projectiles/arrow.java.ftl"/>
+<#if (input$knockback == "/*@int*/0") && (input$piercing == "/*@int*/0")>
+	initArrowProjectile(new ${generator.map(field$projectile, "projectiles", 0)}(${generator.map(field$projectile, "projectiles", 1)}, projectileLevel),
+		${input$shooter}, ${opt.toFloat(input$damage)}, ${field$projectile?starts_with("CUSTOM:")}, ${field$fire == "TRUE"}, ${field$particles == "TRUE"},
+		AbstractArrow.Pickup.${field$pickup})
+<#elseif field$projectile?starts_with("CUSTOM:")>
+	<@addTemplate file="utils/projectiles/arrow_weapon.java.ftl"/>
+	initArrowProjectile(new ${generator.map(field$projectile, "projectiles", 0)}(${generator.map(field$projectile, "projectiles", 1)}, 0, 0, 0, projectileLevel,
+		createArrowWeaponItemStack(projectileLevel, ${opt.toInt(input$knockback)}, (byte) ${input$piercing})), ${input$shooter}, ${opt.toFloat(input$damage)},
+		${field$projectile?starts_with("CUSTOM:")}, ${field$fire == "TRUE"}, ${field$particles == "TRUE"}, AbstractArrow.Pickup.${field$pickup})

```

Comment:
Your outer condition checks if projectile is CUSTOM, so does this repeated check makes sense.

Similar for the second if branch below.

Also applied to 1.21.4

Reply:
:heavy_check_mark: Fixed